File: ScannerPT.java
import java.util.*;
import java.util.regex.*;

public class ScannerPT {

    private static final String TOK_REGEX =
            "(?<RETICENCIAS>\\.\\.\\.)"
            + "|(?<TRACO>--|—)"
            + "|(?<PALAVRA>[A-Za-zÀ-ÖØ-öø-ÿ]+(?:-[A-Za-zÀ-ÖØ-öø-ÿ]+)*)"
            + "|(?<NUMERO>\\d+(?:[.,]\\d+)?)"
            + "|(?<OPERADOR>==|!=|<=|>=|[=+\\-*/<>])"
            + "|(?<PONTUACAO>[.,;:!?()\"'“”‘’\\[\\]{}])"
            + "|(?<ESPACO>\\s+)"
            + "|(?<OUTRO>.)";

    private static final Set<String> ARTIGOS = new HashSet<>(Arrays.asList(
            "o", "a", "os", "as", "um", "uma", "uns", "umas"
    ));

    private static final Set<String> PRONOMES = new HashSet<>(Arrays.asList(
            "eu", "tu", "ele", "ela", "nós", "nos", "vós", "vos",
            "eles", "elas", "me", "te", "se", "lhe", "lhes", "mim",
            "comigo", "ti", "contigo", "si", "consigo", "isto", "isso",
            "aquilo", "este", "esta", "esse", "essa", "aquele", "aquela",
            "quem", "que", "qual", "quais"
    ));

    private static final Set<String> PREPOSICOES = new HashSet<>(Arrays.asList(
            "de", "do", "da", "dos", "das", "em", "no", "na", "nos", "nas",
            "por", "para", "a", "ao", "à", "às", "até", "com", "sem", "sob",
            "sobre", "entre"
    ));

    private static final Set<String> CONJUNCOES = new HashSet<>(Arrays.asList(
            "e", "ou", "mas", "porém", "todavia", "contudo", "porque", "pois",
            "embora", "quando", "se"
    ));

    public static class Token {
        public final String lexema;
        public final String tipo;

        public Token(String lexema, String tipo) {
            this.lexema = lexema;
            this.tipo = tipo;
        }

        @Override
        public String toString() {
            return "(" + lexema + ", " + tipo + ")";
        }

        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof Token)) return false;
            Token other = (Token) obj;
            return lexema.equals(other.lexema) && tipo.equals(other.tipo);
        }

        @Override
        public int hashCode() {
            return Objects.hash(lexema, tipo);
        }
    }

    public static String classificarPalavra(String palavra) {
        String p = palavra.toLowerCase();

        if (ARTIGOS.contains(p)) return "ARTIGO";
        if (PRONOMES.contains(p)) return "PRONOME";
        if (PREPOSICOES.contains(p)) return "PREPOSICAO";
        if (CONJUNCOES.contains(p)) return "CONJUNCAO";
        if (p.endsWith("ar") || p.endsWith("er") || p.endsWith("ir")
                || p.endsWith("ando") || p.endsWith("endo") || p.endsWith("indo")
                || p.endsWith("ava") || p.endsWith("iam")) {
            return "VERBO?";
        }
        if (p.endsWith("mente")) return "ADV?";
        return "PALAVRA";
    }

    public static List<Token> tokenize(String texto) {
        List<Token> tokens = new ArrayList<>();
        Pattern pattern = Pattern.compile(TOK_REGEX);
        Matcher matcher = pattern.matcher(texto);

        while (matcher.find()) {
            if (matcher.group("ESPACO") != null) {
                continue;
            } else if (matcher.group("PALAVRA") != null) {
                String lexema = matcher.group("PALAVRA");
                tokens.add(new Token(lexema, classificarPalavra(lexema)));
            } else if (matcher.group("NUMERO") != null) {
                tokens.add(new Token(matcher.group("NUMERO"), "NUMERO"));
            } else if (matcher.group("OPERADOR") != null) {
                tokens.add(new Token(matcher.group("OPERADOR"), "OPERADOR"));
            } else if (matcher.group("PONTUACAO") != null) {
                tokens.add(new Token(matcher.group("PONTUACAO"), "PONTUACAO"));
            } else if (matcher.group("TRACO") != null) {
                tokens.add(new Token(matcher.group("TRACO"), "TRACO"));
            } else if (matcher.group("RETICENCIAS") != null) {
                tokens.add(new Token(matcher.group("RETICENCIAS"), "RETICENCIAS"));
            } else {
                tokens.add(new Token(matcher.group(), "??"));
            }
        }

        return tokens;
    }

    public static List<String> tokenizeToStrings(String texto) {
        List<String> saida = new ArrayList<>();
        for (Token t : tokenize(texto)) {
            saida.add(t.lexema);
        }
        return saida;
    }

    public static void main(String[] args) {
        String exemplo = "Do título. Uma noite destas, vindo da cidade para o Engenho Novo.";
        List<Token> tokens = tokenize(exemplo);
        for (Token t : tokens) {
            System.out.println(t);
        }
    }
}

File: scanner_pt.py
import re
from typing import List, Tuple

Token = Tuple[str, str]

TOKEN_SPECIFICATION = [
    ("RETICENCIAS", r"\.\.\."),
    ("TRACO", r"--|—"),
    ("PALAVRA", r"[A-Za-zÀ-ÖØ-öø-ÿ]+(?:-[A-Za-zÀ-ÖØ-öø-ÿ]+)*"),
    ("NUMERO", r"\d+(?:[.,]\d+)?"),
    ("OPERADOR", r"==|!=|<=|>=|[=+\-*/<>]"),
    ("PONTUACAO", r"[.,;:!?()\"'“”‘’\[\]{}]"),
    ("ESPACO", r"\s+"),
    ("OUTRO", r".")
]

TOK_REGEX = "|".join(f"(?P<{name}>{pattern})" for name, pattern in TOKEN_SPECIFICATION)

ARTIGOS = {
    "o", "a", "os", "as", "um", "uma", "uns", "umas"
}

PRONOMES = {
    "eu", "tu", "ele", "ela", "nós", "nos", "vós", "vos",
    "eles", "elas", "me", "te", "se", "lhe", "lhes", "mim",
    "comigo", "ti", "contigo", "si", "consigo", "isto", "isso",
    "aquilo", "este", "esta", "esse", "essa", "aquele", "aquela",
    "quem", "que", "qual", "quais"
}

PREPOSICOES = {
    "de", "do", "da", "dos", "das", "em", "no", "na", "nos", "nas",
    "por", "para", "a", "ao", "à", "às", "até", "com", "sem", "sob",
    "sobre", "entre"
}

CONJUNCOES = {
    "e", "ou", "mas", "porém", "todavia", "contudo", "porque", "pois",
    "embora", "quando", "se"
}


def classificar_palavra(palavra: str) -> str:
    p = palavra.lower()

    if p in ARTIGOS:
        return "ARTIGO"
    if p in PRONOMES:
        return "PRONOME"
    if p in PREPOSICOES:
        return "PREPOSICAO"
    if p in CONJUNCOES:
        return "CONJUNCAO"
    if p.endswith(("ar", "er", "ir", "ando", "endo", "indo", "ava", "iam")):
        return "VERBO?"
    if p.endswith("mente"):
        return "ADV?"
    return "PALAVRA"


def tokenize(texto: str) -> List[Token]:
    tokens: List[Token] = []

    for mo in re.finditer(TOK_REGEX, texto):
        tipo = mo.lastgroup
        lexema = mo.group()

        if tipo == "ESPACO":
            continue
        if tipo == "PALAVRA":
            tokens.append((lexema, classificar_palavra(lexema)))
        elif tipo == "NUMERO":
            tokens.append((lexema, "NUMERO"))
        elif tipo == "OPERADOR":
            tokens.append((lexema, "OPERADOR"))
        elif tipo == "PONTUACAO":
            tokens.append((lexema, "PONTUACAO"))
        elif tipo == "TRACO":
            tokens.append((lexema, "TRACO"))
        elif tipo == "RETICENCIAS":
            tokens.append((lexema, "RETICENCIAS"))
        else:
            tokens.append((lexema, "??"))

    return tokens


def tokenize_to_strings(texto: str) -> List[str]:
    return [lexema for lexema, _ in tokenize(texto)]


def processar_arquivo(caminho_entrada: str, caminho_saida: str) -> None:
    with open(caminho_entrada, "r", encoding="utf-8") as f:
        conteudo = f.read()

    tokens = tokenize_to_strings(conteudo)

    with open(caminho_saida, "w", encoding="utf-8") as f:
        f.write(str(tokens))


if __name__ == "__main__":
    exemplo = 'Do título. Uma noite destas, vindo da cidade para o Engenho Novo, encontrei no trem.'
    resultado = tokenize(exemplo)

    for token in resultado:
        print(token)
